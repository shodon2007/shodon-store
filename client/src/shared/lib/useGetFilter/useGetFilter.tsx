import { useQuery } from "@tanstack/react-query";
import filterApi from "src/shared/api/filterApi";

const useGetFilter = (type: string) => {
  const query = useQuery({
    queryKey: ["filter", type],
    queryFn: async () => {
      const res = await filterApi.getFilter(type);
      return res.data;
    },
  });

  return query;
};

export { useGetFilter };
