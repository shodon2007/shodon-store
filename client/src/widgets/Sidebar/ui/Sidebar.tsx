import { Dispatch, FC, SetStateAction, useEffect } from 'react';
import { useParams, useSearchParams } from 'react-router-dom';

import { FilterType } from 'src/shared/api/filterApi';
import { Attribute } from 'src/shared/api/productApi';
import classNames from 'src/shared/lib/classNames/classNames';
import { useGetFilter } from 'src/shared/lib/useGetFilter/useGetFilter';

import toggleAttribute from '../../../features/toggleFilterButton/model/toggleAttribute';

import cls from './Sidebar.module.scss';
import checkFilterStatus from '../model/checkFilterStatus';
import ToggleFilterButton from 'src/features/toggleFilterButton/ui/toggleFilterButton';

interface FilterProps {
	className?: string;
	filters: FilterType;
	setFilters: Dispatch<SetStateAction<FilterType>>;
}

const Sidebar: FC<FilterProps> = ({ filters, setFilters }) => {
	const [, setSearchParams] = useSearchParams();
	const { type } = useParams();

	const { data: attributes, isLoading } = useGetFilter(type);

	useEffect(() => {
		setSearchParams(filters);
	}, [filters]);

	if (isLoading) {
		return <div>Загрузка фильтров...</div>;
	}

	function toggleAttributeHandler(attribute: Attribute, checked: boolean) {
		setFilters(prev => toggleAttribute(prev, attribute, checked));
	}

	return (
		<div className={classNames(cls.Sidebar, {}, [])}>
			{Object.entries(attributes).map(([title, descriptions], index) => {
				return (
					<div key={index}>
						<span>{title}</span>
						<div className={cls.attributes}>
							{descriptions.map((description, index) => {
								const checked = checkFilterStatus(filters, title, description);

								return (
									<ToggleFilterButton
										checked={checked}
										description={description}
										title={title}
										toggleAttributeHandler={toggleAttributeHandler}
										key={index}
									/>
								);
							})}
						</div>
					</div>
				);
			})}
		</div>
	);
};
export default Sidebar;
