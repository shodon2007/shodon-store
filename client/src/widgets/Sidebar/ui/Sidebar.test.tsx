import { render, screen } from '@testing-library/react';
import { useGetFilter } from 'src/shared/lib/useGetFilter/useGetFilter';
import Sidebar from './Sidebar';
import { MemoryRouter } from 'react-router-dom';

jest.mock('src/shared/lib/useGetFilter/useGetFilter');

describe('testing sidebar', () => {
	test('testing attributes', () => {
		(useGetFilter as jest.Mock).mockReturnValue({
			data: {
				NFC: ['ЕСТЬ', 'НЕТ'],
				Камера: ['8mp', '12mp'],
				Озу: ['6gb', '8gb'],
			},
			isLoading: false,
		});
		render(
			<MemoryRouter initialEntries={['']}>
				<Sidebar filters={{}} setFilters={() => false} />
			</MemoryRouter>,
		);

		const attributeButtons = screen.getAllByTestId('attribute-button');

		expect(screen.getByTestId('sidebar')).toBeInTheDocument();
		expect(screen.getByTestId('sidebar')).toMatchSnapshot();

		expect(attributeButtons.length).toBe(6);
	});
});
