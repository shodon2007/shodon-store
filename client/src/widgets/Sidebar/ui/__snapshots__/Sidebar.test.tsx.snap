// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing sidebar testing attributes 1`] = `
<div
  class="Sidebar"
  data-testid="sidebar"
>
  <h1
    class="title"
  >
    Сортировка
  </h1>
  <select
    name="dropdown"
  >
    <option
      value="datedesc"
    >
      Сначала новые
    </option>
    <option
      value="dateasc"
    >
      Сначала старые
    </option>
    <option
      value="pricedesc"
    >
      По убыванию цены
    </option>
    <option
      value="priceasc"
    >
      По возрастанию цены
    </option>
    <option
      value="name"
    >
      По названию
    </option>
  </select>
  <button
    class="Button attributesBtn"
  >
    <div />
  </button>
  <div />
  <div
    class="attributeList attributesDesctop"
  >
    <div>
      <span
        class="attributeTitle"
      >
        NFC
      </span>
      <div
        class="attributes"
      >
        <button
          class="Button smallest attribute"
          data-testid="attribute-button"
        >
          <span>
            ЕСТЬ
          </span>
        </button>
        <button
          class="Button smallest attribute"
          data-testid="attribute-button"
        >
          <span>
            НЕТ
          </span>
        </button>
      </div>
    </div>
    <div>
      <span
        class="attributeTitle"
      >
        Камера
      </span>
      <div
        class="attributes"
      >
        <button
          class="Button smallest attribute"
          data-testid="attribute-button"
        >
          <span>
            8mp
          </span>
        </button>
        <button
          class="Button smallest attribute"
          data-testid="attribute-button"
        >
          <span>
            12mp
          </span>
        </button>
      </div>
    </div>
    <div>
      <span
        class="attributeTitle"
      >
        Озу
      </span>
      <div
        class="attributes"
      >
        <button
          class="Button smallest attribute"
          data-testid="attribute-button"
        >
          <span>
            6gb
          </span>
        </button>
        <button
          class="Button smallest attribute"
          data-testid="attribute-button"
        >
          <span>
            8gb
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
